---
title: "R Notebook"
output: html_notebook
---


Link transcriptome reads from GLAMR to working group folder
```{r}

# Find reads in GLAMR and make column with a path to where they should be linked in the working group directory
read_df <- Sys.glob("/geomicro/data2/kiledal/GLAMR/data/projects/set_62/metatranscriptomes/*/reads/decon_*_reads_fastp.fastq.gz") %>% 
  data.frame(path = .) %>% 
  unglue::unglue_unnest(path, "/geomicro/data2/kiledal/GLAMR/data/projects/{StudyID}/metatranscriptomes/{SampleID}/reads/decon_{read_direction}_reads_fastp.fastq.gz",remove = FALSE) %>% 
  mutate(link_path = str_glue("/geomicro/data2/kiledal/projects/2024_working_group/data/transcriptomes/{SampleID}_{read_direction}.fastq.gz"))

# Make directory to link reads into
read_df$link_path %>% dirname() %>% unique() %>% fs::dir_create()

# Link the reads
fs::link_create(read_df$path, read_df$link_path)

```


```{r}
header = c("id",	"type",	"start",	"stop", "strand", "locus_tag",	"gene", "product", "x_refs")

gtf_file <- read_tsv("/geomicro/data2/kiledal/projects/2024_working_group/data/reference/genomes/GCF_002095975/GCF_002095975.tsv",skip = 5) %>% #col_names = header)
  janitor::clean_names()

annotations <- gtf_file %>% 
  #separate_wider_delim(x_refs, delim = ";", names= c("ID","Name","locus_tag","product","DBxref"),too_few = "align_start",too_many = "drop",names_sep = "|") #%>% 
    select(GeneID = "locus_tag",Chr = "number_sequence_id",Start = "start",End = "stop", Strand = "strand") %>% 
    filter(!is.na(GeneID))

mapping_counts <- Rsubread::featureCounts(files = "/geomicro/data2/kiledal/projects/2024_working_group/data/transcriptome_bams/GCF_002095975/samp_4514.bam",annot.ext = annotations
                                          #annot.ext = "/geomicro/data2/kiledal/projects/2024_working_group/data/reference/genomes/GCF_002095975/GCF_002095975.gff3",
                                          #isGTFAnnotationFile = TRUE,
                                          #GTF.featureType = "locus_tag"
                                          )

mapping_counts_table <- mapping_counts$counts %>% 
  as.data.frame()

```




